Enhance cibuildwheel workflow

AllSeeingEyeTolledEweSew opened the pull request:
This adds various enhancements to the `cibuildwheel` workflow, in hopes of making it align well with release and build processes.

I can change or split up this PR as you like. My intent is to throw a lot of enhancements at the wall, and see which ones you'd like to keep.

## ~~`master` broken?~~

~~**NOTE** that the `cibuildwheel` jobs currently fail because `tests.py` segfaults on every platform.~~

~~I tried rebasing this change on ab4c2a52006b90f58922762224f1e5e6f370b746, and the tests passed then, on every platform. I haven't done this rebase in this PR, due to the conflict with the existing `build.yml`.~~

~~I presume this means it would actually be useful to have automatic building and testing on every PR, so I restored the PR trigger for this workflow.~~

## Changes in this PR

* **Added `setup.py configure_pinned`**: modifies `setup.cfg` in-place to pin the sqlite version
* **Move `cibuildwheel` config into `pyproject.toml`**
  * Makes it easier to do local test builds, getting consistent config
* ~~**Default to building against the "mached" version of sqlite**~~
  * ~~so `apsw-3.36.0` builds against sqlite 3.36.0), per [this comment](https://github.com/rogerbinns/apsw/issues/310#issuecomment-976239211). This guarantee will make it more straightforward for apps to declare a dependency on a particular version of sqlite/apsw.~~
  * ~~The current behavior of using the latest sqlite is available using `setup.py fetch --version=latest`~~
* **Workflow *does not* automatically run on new git tags**
  * Since the build may fail, I presume it's nicer to be able to (re-)run the build at will, without changing git tags.
* **Workflow can be manually run**
  * When manually run, we will build wheels in the full set of flavors supported by `cibuildwheel`.
  * This configuration splits the work into roughly-equal-time jobs. Building all flavors takes about an hour, as the emulated architectures take a long time.
* **Optional uploading to PyPI**
  * This option is only available when the workflow is manually run.
  * PyPI API keys are stored as github secrets.
* ~~**Workflow will automatically run for each PR**~~
  * ~~In [the original PR](https://github.com/rogerbinns/apsw/pull/313) this was a non-feature. However the build is currently broken in master so I presume this would be useful!~~
  * ~~When triggered by a PR, we will do a limited selection of build flavors, not the full selection, for time's sake.~~
* **Build artifacts are cached for fast re-runs**
  * You can run several build workflows for testing, then re-run the last one with PyPI upload enabled.
  * This plus `fail-fast: false` is also helpful to recover from spurious failures (e.g. failing to fetch sqlite source). If one job fails, just re-run the whole workflow, and any successful artifacts will be picked up from the cache on the second run.
* ~~**Automatically update `cibuildwheel` using `dependabot`**~~
  * ~~Per https://cibuildwheel.readthedocs.io/en/stable/faq/#option-1-github-action~~

## Proposed release procedure

1. Run the build workflow, *without* PyPI upload
2. If all build flavors succeed, add a git tag
3. Re-run the build workflow *with* PyPI upload. The build artifacts will be cached, so the re-run will be fast

## Example run

~~[This workflow run](https://github.com/AllSeeingEyeTolledEweSew/apsw/actions/runs/1498135732) shows all build flavors. I ran this with the change rebased on ab4c2a52006b90f58922762224f1e5e6f370b746, so the tests would pass. I enabled PyPI upload for demonstration purposes, which obviously fails for lack of the API secret.~~

## Optional further enhancements

* **Drop support for `ppc64le` and `s390x`**
  * These platforms are rare, and you may want to wait until some user actually requests support for them before spending time and storage on them.
  * I think `aarch64` is worth keeping, as I believe this covers modern raspberry pi and linux/docker on M1 Macs.
* **Native artifacts for M1 Macs**
  * `cibuildwheel` can cross-compile these wheels, but they cannot be tested. See [this FAQ entry](https://cibuildwheel.readthedocs.io/en/stable/faq/#apple-silicon). Github Actions doesn't have M1 mac runners yet (see https://github.com/actions/runner/issues/805). AFAICT everyone building native wheels for M1 macs are using self-hosted CI solutions.

rogerbinns commented:
Are different goals being muddled up here?

The only thing I need is a (preferably manual) push button that builds source and binary wheels and sends to pypi.  I already do all my testing locally.

Are you also trying to make this useful for folks doing pull requests and similar contributions?  If so then great, but I have no feedback since I don't do that!

I definitely won't change setup.py to pin the SQLite version.  It can be pinned when making the  source distribution or similar (eg command becomes ````... sdist --pin-sqlite=1.2.3```` and alters files as part of that process).

It is possible to build the test extension and run the tests with setup.py args: ````python setup.py build --enable-all-extensions build_test_extension test````

Once 3.37 is out things can be simplified.

AllSeeingEyeTolledEweSew commented:
> Are different goals being muddled up here?

My one goal is to get distributions of apsw into pypi, and maintained there. I wanted to streamline as much as possible to make that easier.

> The only thing I need is a (preferably manual) push button that builds source and binary wheels and sends to pypi. I already do all my testing locally.

If you mean my comment regarding `pyproject.toml`, I meant this makes it easier to locally test the cibuildwheel setup (on Linux at least, see [here](https://cibuildwheel.readthedocs.io/en/stable/options/#platform)), not locally test apsw. The alternative to having cibuildwheel config in `pyproject.toml` is to set a bunch of environment variables.

> Are you also trying to make this useful for folks doing pull requests and similar contributions? If so then great, but I have no feedback since I don't do that!

If you mean that I re-enabled the pull request trigger for `build.yml`, I actually thought this might be more useful for your own development than for contributions. Since the tests currently break in `master` as I mentioned, I assume some additional automated testing (i.e. using the workflow of creating your own pull requests against apsw) could have prevented that.

> I definitely won't change setup.py to pin the SQLite version. It can be pinned when making the source distribution or similar (eg command becomes `... sdist --pin-sqlite=1.2.3` and alters files as part of that process).

I would find this extremely surprising as a user. I always expect that if I install an sdist, I'm just running `setup.py` from a given version in source control.

I understand that pinning the sqlite version is itself a surprising change to existing users, but I imagined that the behavior change (of pinning the sqlite version) across apsw versions was less surprising to fewer people than the sdist always being different from the contents of source control.

(I mistook [this comment](https://github.com/rogerbinns/apsw/issues/310#issuecomment-977024899) to mean you *did* want to pin sqlite version going forward, which is why I made this change)

AllSeeingEyeTolledEweSew commented:
Adding to my comments about surprise: I already found it extremely surprising that apsw did not already pin its build to a given sqlite version, since its design goal is to expose features of a given version of sqlite, and its version scheme follows sqlite's.

rogerbinns commented:
Last comment first: You missed one small important piece :-)  SQLite retains backwards compatibility - ie a piece of code written against the very first release of SQLite 3 (2004) remains both API (source) and binary compatible with all subsequent SQLite 3 releases.  They even put effort into ensuring query plans remain the same or better - ie if you upgrade SQLite only, you won't suddenly find a previously quick query going slower.

So yes APSW does target a particular SQLite version, but importantly it is that SQLite version **or later**.  You could have pinned a version of APSW from 10 years ago and still keep updating SQLite and everything would be fine.  This thinking is also part of why APSW supports Python versions going back ~20 years.  You can use any Python version, APSW version, and SQLite version, **and stuff keeps working**.  (I also made the decision to not support older SQLite with newer APSW because that makes testing and coding way more complicated, hence the tying of APSW and SQLite version numbers.)

AllSeeingEyeTolledEweSew commented:
> SQLite retains backwards compatibility

True, but they do introduce behavior changes.

Let's say I have some query that utilizes (without loss of generality) the "ESCAPE clause of a LIKE operator now overrides wildcard characters" behavior introduced in 3.32.0. If I am using apsw < 3.32.0, I *may or may not* get this behavior. API/ABI compatibility doesn't help me; I still need to know and/or control the version.

> So yes APSW does target a particular SQLite version, but importantly it is that SQLite version **or later**

Is that noted in documentation anywhere? I couldn't find it.

I suppose for my example above, you'd say that I should either
* declare a dependency on `apsw>=X.Y` and rely on the fact that this targets sqlite **>=** X.Y
* check `sqlitelibversion()`, when the above is infeasible (e.g. middleware code)

I do understand your goal of having a "floating" sqlite version. I still think it's surprising and disagree with it, but I expect I'm unlikely to convince you without some objective data, which I don't have.

That being the case, then IMO it's better for wheels and sdists on pypi to **not** pin their sqlite version. As much as the floating version is surprising, I think it's more surprising if pypi has different behavior from what's in source control.

rogerbinns commented:
Earlier comment now: there are tiers of code

git/master: this is where development takes place and it is **not** a source distribution.  Only developers working on APSW itself (ie me) use this.  It does indeed currently segfault due to issues in the cyclic gc code.  I still need to find out if tp_clear has to be implemented, amongst other things.  Unfortunately I can't stop github from labeling this as source.  Yes I could use branches but that makes things more complicated, and doesn't benefit me or any APSW users.

Official releases (source): Produced [from master](https://github.com/rogerbinns/apsw/wiki/Release-Procedure) but especially important that it defaults to the current SQLite version, and has command line flags to control extensions, downloading or using local SQLite etc.  This is what a project using SQLite should use so they can configure things as they want, and get no surprises now or in the future.

Official releases (windows binaries): Provided because compiling on Windows is a pain, especially if using an older Python version.

To which we (mostly you :-)) are adding pypi source and binary releases.  This is for the crowd who just want something to work, don't want to compile themselves if possible, and don't want to think about future APSW and SQLite versions, which extensions are in or out, or anything similar.  They need the least surprises, and it is best that if you install on Monday and on Thursday you get the same results, even if there was a SQLite release inbetween.

I'd expect a [Makefile](https://github.com/rogerbinns/apsw/blob/master/Makefile#L187) target like ````source_for_pypi```` that can produce a source archive suitably configured.  It may be possible to pin the SQLite version in the setup.cfg.

Pull requests haven't been helpful to me.  In general I am wary of including code from others because that spreads the copyright.  They also are usually missing documentation updates and test cases.  Those two pieces take longer to write, and are often several times larger than the pull request itself.  So I usually use the pull request as inspiration, rather than mergeable code.

That said I do intend to merge your work, and greatly appreciate it.

rogerbinns commented:
The version compatibility is the very first thing in [the doc](https://rogerbinns.github.io/apsw/).  Right now it says:

    Use with SQLite **3.36 or later**, CPython (2.3 and later, 3.1 and later).

It isn't always the case that the SQLite version is bumped.  For example if no constants are added or APSW using new APIs then the minimum SQLite version is left unchanged.  You can see how APSW 3.35 worked with SQLite 3.34 and above [here](https://web.archive.org/web/20210516183239/https://rogerbinns.github.io/apsw/index.html).

You are correct that SQLite introduces behaviour changes, but they are opt in unless bug fixes.  But do understand that APSW does not alter the behaviour of SQLite!  It really is the thinnest layer to communicate between Python and SQLite's C API.  With your example of a change, you should make a conscious decision about versions  which is [precisely what the existing non-pypi source distribution does](https://rogerbinns.github.io/apsw/build.html#setup-py).  It was the [historic inability](https://rogerbinns.github.io/apsw/download.html#easy-install-pip-pypi) to specify options when using pip/pypi that led me to not publishing there.

I still think it is best that official source (non-pypi) default to latest SQLite since it is trivial to control that and the other build flags.  And that pypi is a pinned SQLite release so that binaries and sdist will give the same result no matter when installed (eg if a new SQLite release happens).

AllSeeingEyeTolledEweSew commented:
> git/master: this is where development takes place and it is **not** a source distribution. Only developers working on APSW itself (ie me) use this. It does indeed currently segfault due to issues in the cyclic gc code.

So IIUC, `master` is *not* expected to be functional / pass its own tests, is that right? If so then I should remove the PR trigger in the workflow, since the point of PR checks is to maintain the invariant that tests always pass

> It isn't always the case that the SQLite version is bumped. For example if no constants are added or APSW using new APIs then the minimum SQLite version is left unchanged. You can see how APSW 3.35 worked with SQLite 3.34 and above [here](https://web.archive.org/web/20210516183239/https://rogerbinns.github.io/apsw/index.html).

This seems to make it hard to declare dependencies.

Let's say sqlite 3.38 introduces new behavior, but does not introduce new constants or APIs. Let's say my app depends on the new behavior. I'd like to write my `setup_requires` to indicate this dependency, the way I do for all other modules. I'd like to write something like `setup_requires=apsw>=3.38`. But then it seems that either
* apsw 3.38 may include sqlite 3.37, and thus cause my app to break at runtime, or
* perhaps there just *wouldn't be* a release of apsw 3.38?

> You are correct that SQLite introduces behaviour changes, but they are opt in unless bug fixes. But do understand that APSW does not alter the behaviour of SQLite! It really is the thinnest layer to communicate between Python and SQLite's C API. With your example of a change, you should make a conscious decision about versions which is [precisely what the existing non-pypi source distribution does](https://rogerbinns.github.io/apsw/build.html#setup-py). It was the [historic inability](https://rogerbinns.github.io/apsw/download.html#easy-install-pip-pypi) to specify options when using pip/pypi that led me to not publishing there.

What do you mean by "a conscious decision about versions"? I'm not understanding what you mean from the documentation link.

> I still think it is best that official source (non-pypi) default to latest SQLite since it is trivial to control that and the other build flags. And that pypi is a pinned SQLite release so that binaries and sdist will give the same result no matter when installed (eg if a new SQLite release happens).

I think I'm having trouble understanding the use case you're designing for, or how common it is.

apsw puts emphasis on customizing the apsw/sqlite build, but who is doing this? Are there lots of users who need apsw with some extensions *disabled*? Are there lots of users who want to run an old apsw with a newer sqlite? What apps require this kind of setup? How are these apps installed?

I'm a fan of disabling unused features (I ran gentoo for a long time), but I'm having trouble understanding why apsw *requires* customization. I can't figure out how to write an app that uses apsw, and *not* require users to run special commands to install apsw "correctly" for my app. Either I just haven't understood how to do it, or I don't understand why the simple case is sacrificed for the sake of customization.

rogerbinns commented:
> So IIUC, master is not expected to be functional / pass its own tests, is that right?

If it doesn't pass its own tests then that is work that is still ongoing - a simple 'TODO'.  Unfortunately some people have scripted just grabbing master at random instead of the source distro or at the very least use tags.  I've deliberately left it broken recently to discourage that behaviour.

> This seems to make it hard to declare dependencies

Yes, which is why it wasn't on pypi!  You either need to provide source or compiled SQLite of your chosen version, specify --fetch of your chosen version, or let fetch do its thing (latest is using source, pinned if using pypi).

> perhaps there just wouldn't be a release of apsw 3.38?

I never guaranteed that each SQLite release would have a corresponding APSW release.  Historically SQLite had a far more variable release schedule and didn't use semver.  So both projects had more random releases.  SQLite did adopt semver and a far less frequent and more predictable release schedule.  There were APSW releases where the changelog is "No APSW changes" done purely so there would be updated Windows binaries.

> What do you mean by "a conscious decision about versions"? I'm not understanding what you mean from the documentation link.

I meant what you discussed in the previous section.  You should choose the SQLite version to use, instead of just hoping defaults work out ok.  It is especially important for a project deploying on multiple systems over a time period.  The doc shows how to specify the version you want, and how pypi makes that difficult.

> I think I'm having trouble understanding the use case you're designing for, or how common it is.

The big issue is that platforms update SQLite with a considerable lag - they can be many months if not years behind.  SQLite itself can also have point releases - eg there may be a SQLite/APSW 3.20.0 released and then a few weeks later SQLite 3.20.1 comes out.  I won't do another 'no change' release of APSW because of that.  (This is why there is also a lag between SQLite and APSW releases - I wait to see if a point release comes out.)

Source (non-pypi) installs have been the norm until this PR is done.  It has defaulted to the latest release, but is trivial to specify a particular SQLite release.  This is how devs who care about SQLite should be doing things.

pypi: Devs are just going to add apsw to requirements and not care.  In order to ensure that source and binary releases are the same the SQLite version will be pinned.  If you install on a Monday and a Friday you'll get the same SQLite version, even if there was a new SQLite release on Wednesday.  This is the "least surprises" approach.

> apsw puts emphasis on customizing the apsw/sqlite build, but who is doing this

Everyone :)  Because platform SQLite versions are so laggy you have to do something to provide newer SQLite.  Extensions are an attack surface and somewhere for unintentional bugs to happen.  Already this PR messes up the ICU extension - enable all extensions only turns it on if the ICU dev packages are present, which they aren't in the ci environment.  And if they were then you'd get dll hell on deployment.

Running older APSW with newer SQLite is just fine - there is no need to upgrade to newer APSW unless you want specific features, or updated SQLite constants etc.  Phrased another way - updating SQLite will never require updating APSW.

I was also a gentoo user for ages!  You can always use the APSW/SQLite provided by your system which will be months to years out of date, and that is fine for many.  But I couldn't solve the problem of pypi and what should happen if there are SQLite releases and no APSW releases over time, and especially because there was no way of telling pip "tell APSW to fetch SQLite v3.4.5".  This PR will fix that.

AllSeeingEyeTolledEweSew commented:
I think we agree now that pypi releases should include a pinned version of sqlite + dependencies, but I want to be sure I understand the details.

In particular I rescind my earlier idea, that pypi releases should include "bounded" sqlite versions (>=X.Y.Z but not necessarily ==X.Y.Z), so that the sdist has the same behavior as source control. Expectations aside, I think the biggest issue with this is that `setup.py` invoked from `pip` can only do its default behavior of trying to use system sqlite headers, and as you mention, these are probably old. I expect in the majority of cases, a python app would require a new-ish apsw, and users wouldn't be able to install a new enough system sqlite to match it.

>  I've deliberately left [master] broken recently

FWIW this makes collaboration really hard. I can't meaningfully test anything. If this PR gets merged now, then I expect when master is later fixed, you'll go to test this workflow and find it broken. If you wait until release time to merge the PR, then it delays testing and could hold up the release.

I find this extraordinarily surprising and I'd expect any modern developer to be equally surprised. This seems an extreme reaction to some users building from `master` and expecting support.

I don't expect you to change your workflow of course, but I think this feedback is warranted.

> I never guaranteed that each SQLite release would have a corresponding APSW release. ... There were APSW releases where the changelog is "No APSW changes" done purely so there would be updated Windows binaries.

> there may be a SQLite/APSW 3.20.0 released and then a few weeks later SQLite 3.20.1 comes out. I won't do another 'no change' release of APSW because of that.

> pypi: ... In order to ensure that source and binary releases are the same the SQLite version will be pinned. If you install on a Monday and a Friday you'll get the same SQLite version, even if there was a new SQLite release on Wednesday. This is the "least surprises" approach.

I'm confused -- if we do pypi releases which bundle a pinned version of sqlite, and sqlite releases a new version 3.38.1 which includes only bugfixes over 3.38.0, do you intend to
* do a (pypi-only?) apsw 3.38.1 release to pick up the new sqlite, or
* not do a new release?

> Already this PR messes up the ICU extension - enable all extensions only turns it on if the ICU dev packages are present, which they aren't in the ci environment. And if they were then you'd get dll hell on deployment.

Ah, I wasn't aware of this. I see there's also some zlib integration, which according to [the compile docs](https://www.sqlite.org/compile.html) is "used by extensions". Does this means zlib is required for any of the extensions included with `--enable-all-extensions`?

Are ICU or zlib support included with the win32 binary releases? Are these dependencies (or could they be) statically linked?

> Running older APSW with newer SQLite is just fine - there is no need to upgrade to newer APSW unless you want specific features, or updated SQLite constants etc

I find this very confusing. I'm trying to understand the deployment scenarios apsw currently supports, but I don't understand how "old apsw + new sqlite" fits in them.

From what I understand, a client of apsw *requires control* of the command line arguments to apsw's `setup.py`. So these are the only workable deployment scenarios I can think of:
* Platform-specific builds that vendor all dependencies (docker, snap, mac apps, enterprise environments, etc) directly invoke `setup.py` per their needs
* Platform-agnostic apps either
  * Require the user manually to manually apsw in their environment, according to the app's documentation, or
  * Require a bespoke installer script to installs apsw "correctly" for the app, possibly requiring more input about whether they want to try to use the system sqlite
* Platform-agnostic middleware that lacks control over its installation process (libraries, WSGI apps, etc) must document its requirements and *require all client apps* to do one of the previous options
* Distro packages rely on the fact that the distro's apsw package will dynamically link to the distro's sqlite
  * These must restrict the version of sqlite *and* apsw, since restricting just apsw isn't always sufficient
  * I don't know how a distro package would depend on sqlite extensions. A gentoo package could require USE flags on apsw, but do other distros just require apsw/sqlite have all extensions enabled?

Either I've misunderstood your support scenarios, or it seems like a lot of clients have a lot of pain. The middleware case is particularly painful, and the reason I had to stop using apsw. Only the gentoo case seems painless.

To clarify, I want to support the scenario where python code can just add `setup_requires = apsw>=X.Y.Z`. This implicitly depends on sqlite >= X.Y.Z. `pip install` can get wheels from pypi, distro metapackaging (debhelper, etc) will automatically do the right thing.

AllSeeingEyeTolledEweSew commented:
Asked another way: what's the use case for *not* pinning the version today? What scenarios does this enable? Having separate versioning guarantees for pypi and non-pypi deployments is going to create maintenance burden and user pain, so I really want to understand *why* they should be different. I don't think I can do a good job on this PR unless I understand this pretty well.

AllSeeingEyeTolledEweSew commented:
I did some updates, based my current understanding of apsw's release and dev needs.
* Removed the PR trigger
* Removed dependabot config (dependabot PRs, like all others, will always fail if `master` stays intentionally broken)
* Added a `setup.py configure_pinned` command. This modifies `setup.cfg` in-place to enable all extensions and pin the sqlite version
  * Once done, `bdist_wheel` and `sdist` without args produce artifacts suitable for pypi. This is what's expected by `cibuildwheel` and other tools.
  * To allow `setup.py fetch` to be triggered via config, I added the option `setup.py build --fetch`, so that it's always fetched before a build. This is janky because the fetch is currently triggered in `finalize_options`, which means a lot of redundant fetches in some scenarios, but it does seem to have correct output
  * This is analogous to the `source_for_pypi` makefile target you suggested. I used python instead to use `configparser`, which lets `configure_pinned` be idempotent and refer to the release version directly (`SQLITEVERSION` from `Makefile` is the *minimum* version, not the *pinned* version, IIUC)
  * I used `--enable-all-extensions --enable=icu`, as I want the build to fail if ICU can't be found
* ICU support builds for musllinux, manylinux and macosx builds
  * I haven't attempted to make ICU work for windows, the build just fails right now
  * I admittedly don't know the implications of this. I am assuming that vendoring all shared libraries (the `cibuildwheel` infrastructure does this) is sufficient, but if any of these libraries e.g. try to load unicode data from the system at runtime, the wheel will probably be broken.
  * If either of the above things is hard to resolve, IMO it's reasonable to ship wheels with the ICU extension disabled. pypi is just not a universal dependency management solution yet as you know, but people don't expect it to be.

AllSeeingEyeTolledEweSew commented:
Mad-science thought: I'm not clear on whether extensions like json1 (or others) can be built as separate .so files, but if so, then it might be possible to put those in separate wheels.

I could imagine using `extras_require`, such that client code could depend on `apsw[json1]>=3.37.0`, which would pull in `apsw-json1-3.37.0` containing just the extension. The client code would need to use `db.loadextension("some_canonical_name")`.

rogerbinns commented:
> I find this extraordinarily surprising and I'd expect any modern developer to be equally surprised. This seems an extreme reaction to some users building from master and expecting support.

It isn't users building from master - it is distributors too.  ie when I do a tagged release, they at some point later grab whatever is in master and use that as though it was the tagged release.

On the collaboration side, the most prevalent (and useful) are feature requests, bug reports, questions on how to do things, questions about the doc and code, and partial diffs.  This is a C based project gluing the C APIs of Python and SQLite together, plus a lot of testing code and documentation system.  It is difficult to contribute to at the code level.  You are the first person in the 18 year history of the project to provide more than one patch.

My attitude is that if someone is collaborating I will then make it easier for them in whatever way is helpful.  But I won't make things less convenient for myself every single day, just in case someone might collaborate.  In any event master now works fine and that gnarly issue around cursors and GC took a long time to identify and fix.

It is possible to build the extensions as shared libraries, but then they have to be manually enabled (call load_extension on each one).  Do note that SQLite doesn't have the usual Unixy attitude of splitting everything into lots of pieces, making everything possible share them, and then updating the one true copy of stuff.  That can be done, but opens the doors to many possible errors.  I'd advise folks who want to do that to use the standard library sqlite3 module instead.

You may note that APSW binary is a single file, ideally including SQLite inside.  The compartmentalization works really well, because you can easily end up with multiple copies of SQLite loaded into the same process - for example on MacOS CoreData usage will load SQLite.  ie a specific goal of APSW is to ensure that things can't go wrong.

I don't particularly like the setup.py command hacking, but nothing simpler seems reasonable.  I'd really like that file to get smaller over time.  Note that I won't be merging this PR until after this next release, at which point it will have my full attention.

AllSeeingEyeTolledEweSew commented:
> It isn't users building from master - it is distributors too. ie when I do a tagged release, they at some point later grab whatever is in master and use that as though it was the tagged release.

By distributors do you mean linux distributions? (I'm still trying to get a handle on supported apsw deployments today). Professional packagers expecting expecting support for building from unreleased upstream code is really surprising.

I conjecture that in the past, some people (users and/or developers writing custom install scripts like I mentioned previously) may have found that the best way to get automatic updates of apsw was to use [the documented pip install instructions](https://rogerbinns.github.io/apsw/download.html#i-really-want-to-use-pip) against `master`, since apsw hasn't had first-party integration with pypi or other version management systems. If apsw is regularly published on pypi, I would expect it will lead to fewer people wanting to touch `master` as their least-bad option.

> I won't make things less convenient for myself every single day, just in case someone might collaborate

You're saying using CI to keep master from breaking is *less* convenient for your workflow? That's extremely surprising. Every other developer I've met in decades of software development has thought it *more* convenient, even in absence of collaboration.

> It is difficult to contribute to at the code level

But isn't the lack of CI a major *cause* of this difficulty?

Again, I don't expect you to change your workflow, but a meaningful contribution requires *understanding* it somewhat. I think you won't find it offensive if I say you do things differently from other developers. I have a feeling I haven't worked through all the relevant surprises yet.

> I don't particularly like the setup.py command hacking, but nothing simpler seems reasonable. I'd really like that file to get smaller over time.

Do you mean the use of `setup.cfg` to modify behavior? Or the fact that it's done via a `setup.py` subcommand? It could be a separate script, if I knew the right place to get the version for pinning.

I'm eager to know your thoughts on vendoring the ICU extension. I don't understand it very well. Should it be left out for now? Is it considered essential for FTS?

rogerbinns commented:
> By distributors do you mean linux distributions?

It isn't the mainstream distributions.  It is build scripts for more minor ones where grabbing from master is an easy default - grab an existing template and change the github url.

> ... other developer I've met in decades of software development ...

Commit was done that "worked" fixing #314.  It was discovered that there was still an issue, so I continued diagnosing the issue (Python was accessing a pointer before the Cursor PyObject).  My choices were:

1: Find and fix the cause, leaving master broken for a few days

2: Backout the change, find the cause and fix, backout the backout, commit the fix

I did number 1.  If anyone needed master working then I would do whatever is convenient for them.  Master is undergoing lots of updates anyway because of the new SQLite release.  It is very rare for master to be broken.

> But isn't the lack of CI a major cause of this difficulty?

Not even close.  A complete change has to include:

* C code that makes the desired change
* Python code that tests the desired change
* Code that deals with all possible error states and other interactions with the Python C API
* Code that deals with all possible error states and other interactions with the SQLite C API
* Code added for fault injection (ie cause actual errors/failures to occur) in C and corresponding Python test code
* Running coverage testing and adding more test code to ensure all paths are run
* Documentation that is complete and cross references as necessary, built and viewed in browser
* Running under valgrind and sanitizers with assertions and debug SQLite
* Updating the changelog

CI deals with a **very** small subset of that.  *Very* few people are going to have the time and experience to do all of that.  I am totally fine with that, which is why I consider small amounts of code, bug reports, suggestions and general questions to be valuable collaboration.  I have to do all the pieces above not done, which is ok.  And for anyone who is trying to do more of the above I will do whatever is helpful for them.

> setup.py command hacking

I'd much rather a flag to sdist than a separate command, like is already done to include doc, but understand the ci tool really not wanting extra flags

> right place to get the version for pinning

There is a global variable named ````version```` in setup.py

> ... thoughts on vendoring the ICU extension ...

I think it has to be left out for the pypi builds.  The odds of the ICU library ABI and datafiles being compatible under a variety of OS versions seems slim, given their release cadence.  While ICU is available on Windows I believe it is rarely used.  It is not needed for FTS.

AllSeeingEyeTolledEweSew commented:
> I'd much rather a flag to sdist than a separate command, like is already done to include doc, but understand the ci tool really not wanting extra flags

I thought having a single place to configure everything for pypi was convenient, too. Rather than doing both of `build_ext --configure-pinned` and `sdist --configure-pinned`, we just do one of `setup.py configure_pinned`

> I think it has to be left out for the pypi builds. The odds of the ICU library ABI and datafiles being compatible under a variety of OS versions seems slim, given their release cadence. While ICU is available on Windows I believe it is rarely used. It is not needed for FTS.

Ah, if ICU needs data files then things are messy.

I removed ICU from the build. I wanted to do something like `--enable-all-extensions --except=icu` to ensure ICU *doesn't* build if it happens to be installed in the environment. Instead I just used `--enable=list-of-everything-except-icu`.

I rebased on current master to show that tests pass. [Here's an example run](https://github.com/AllSeeingEyeTolledEweSew/apsw/runs/4453121756?check_suite_focus=true)

I'm unclear if just invoking `tests.py` is appropriate for testing. It looks like there are a lot of paths to invoke it and it responds to a lot of environment variables.

Also, are you aware of this?

```
 /usr/local/lib/python3.9/site-packages/setuptools/dist.py:487: UserWarning: Normalizing '3.37.0-r1' to '3.37.0.post1'
```

`setuptools` has been pushing [PEP 440](https://www.python.org/dev/peps/pep-0440/) normalization since its inception, and looks like it started enforcing it in 46.1.0 (2020-03-21).  `distutils` doesn't seem to normalize, but if you want to rely on it, note that the current trick of using `APSW_FORCE_DISTUTILS` won't be effective for sdists. It looks like `packaging` is removing already-deprecated support for "legacy" version numbers (https://github.com/pypa/packaging/pull/407), though `pip` still supports them for now (https://github.com/pypa/pip/pull/10117) The approaches I see are
* **Ignore normalization**: allow all pypi releases to be `.postN`
  * I expect that once you start releasing to pypi, there will be an increase in confused users asking about it
  * I could see weird stuff happening where e.g. `apsw-3.37.0-r1` is installed in the system, but `pip` is asked to install it in a virtualenv, so it compares `apsw-3.37.0.post2` with `apsw-3.37.0-r1`
* **Bypass normalization**: enforce pypi releases to be `-r1`
  * You'd need some way to modify `setup.py` to be suitable for sdists, I don't think `setup.cfg` will be sufficient
  * I'm not clear on how `pip` compares legacy versions. One piece of code seems to use string sorting, so maybe `3.37.0-r2` would be considered "newer" than `3.37.0-r10`
* **Adopt PEP 440**.
  * I think the best approach is such that `apsw-3.37.0.0` is the first release for a given sqlite version, and `apsw-3.37.0.1` is the second. The `.postN` suffix is best used for packaging issues: pypi treats uploads as absolutely immutable, and `.postN` is the lowest possible significance in PEP 440, so if there's e.g. a packaging issue with `apsw-123.post1` that doesn't require a new code release, the *only* option is to upload `apsw-123.post2`.
  * FWIW, it looks like [alpine at least already mangles the apsw version](https://pkgs.alpinelinux.org/packages?name=*apsw*), though it's only in testing.

I think this is all I can do for now. Let me know if you encounter any issues when integrating this.

rogerbinns commented:
I've updated the doc to clarify why the versions matter and when to use APSW in https://github.com/rogerbinns/apsw/commit/8da7afaedc6fa443741b3fb54125eb5d40a0b82a - feedback appreciated.

If you enable all extensions then ICU is only included if icu-config (older systems) or pkg-config (newer systems) are present with compile time ICU information.  I'd guess the build systems don't have it by default.

> I'm unclear if just invoking tests.py is appropriate for testing.

It is and is what I use.  As you noted it does have extra stuff, but that requires different compilation of APSW (eg enabling fault injection) and tools (like valgrind and coverage).  The useful environment variable is APSWTESTPREFIX which is prepended to all database filenames - I use it to point to tmpfs - as the constant fsyncs SQLite does are instantaneous there versus taking a while on real persistent storage.

I am aware of the version number mangling.  My intention was to do nothing about it unless something breaks, since it only affects filenames.  At that point I'll figure out a better solution.

$APSW_FORCE_DISTUTILS is useful for the older Python versions and has no effect on recent versions.  Remember that this current release supports all versions of Python back to Python 2.3.

AllSeeingEyeTolledEweSew commented:
> I've updated the doc to clarify why the versions matter and when to use APSW in [8da7afa](https://github.com/rogerbinns/apsw/commit/8da7afaedc6fa443741b3fb54125eb5d40a0b82a) - feedback appreciated.

Looks good! I suppose a description of pypi version pinning will come later.

> If you enable all extensions then ICU is only included if icu-config (older systems) or pkg-config (newer systems) are present with compile time ICU information. I'd guess the build systems don't have it by default.

The `cibuildwheel` environments don't include ICU dev tools by default. However, a user may have them when installing an sdist from pypi. I thought it best if we *don't* enable ICU even in this case, so that the pypi releases have a consistent feature set in all environments

rogerbinns commented:
The plan is to finish the release in progress (3.37.0), and then come back this.

The distribution goal is:

* Using the source dist from the download page is recommended and you can choose functionality and versions as appropriate

* Using source dist or binary dist from pypi gets you the same results no matter when you install it, subsequent SQLite releases etc.  People using ICU should really do the first one, so we'll try to exclude it when installing from pypi

rogerbinns commented:
Thank you so much for doing this work.  The result now lives as [build.yml](https://github.com/rogerbinns/apsw/blob/master/.github/workflows/build.yml) and does a full publish to [test pypi](https://test.pypi.org/project/apsw/).  The discussion above helped change defaults for fetch, and made sure all the angles are covered for a robust deployment.  The next (imminent) APSW release will be on the real pypi.
